### PR TITLE
Add debug_name attribute to JIT

### PR DESCRIPTION
**Pull request by BacharL**
Summary:
To enable name attribute to be used for debugging purposes.
Subgraph matcher can be asked to ignore the debug name attribute when comparing graphs.
Rewriter will name the inserted graph nodes with the prefix of the insertion point name.

Fixes #ISSUE_NUMBER

### CODE REVIEW

**Comment by hablb** [comment]
@eellison 
Hi. Can you please review this patch?
Regards.

File: torch/csrc/jit/ir/ir.cpp [hunk 1]
```diff
@@ -2230,8 +2230,19 @@ std::vector<Value*> insertGraph(
   for (const auto i : c10::irange(inputs.size())) {
     value_map[callee.inputs()[i]] = inputs[i];
   }
+
+  //set new nodes deubg_name to prefix to insertion point debug_name
```
Comment (wconstab) [inline torch/csrc/jit/ir/ir.cpp:2234]: typo

**Comment by wconstab** [comment]
Can you give a quick example of how this would be used?

File: torch/csrc/jit/ir/ir.cpp (hunk 1 above)
Comment (hablb) [reply torch/csrc/jit/ir/ir.cpp]: Done

**Comment by hablb** [comment]
> Can you give a quick example of how this would be used?

```
import torch

class MyCell(torch.nn.Module):
    def __init__(self):
        super(MyCell, self).__init__()
        self.linear = torch.nn.Linear(4, 4)

    def forward(self, x, h):
        new_h = torch.tanh(self.linear(x) + h)
        return new_h, new_h

my_cell = MyCell()
x, h = torch.rand(3, 4), torch.rand(3, 4)
traced_cell = torch.jit.trace(my_cell, (x, h))

#find aten::add node
graph = traced_cell.graph.nodes()
node_iter = next(graph)
while node_iter.kind() != 'aten::add':
    node_iter = next(graph)

#set debug_name attribute in jit graph
node_iter.s_('debug_name','MyCell/add1')

print(traced_cell)
#graph(%self.1 : __torch__.MyCell,
#      %x : Float(3, 4, strides=[4, 1], requires_grad=0, device=cpu),
#      %h : Float(3, 4, strides=[4, 1], requires_grad=0, device=cpu)):
#  %linear : __torch__.torch.nn.modules.linear.Linear = prim::GetAttr[name="linear"](%self.1)
#  %20 : Tensor = prim::CallMethod[name="forward"](%linear, %x)
#  %11 : int = prim::Constant[value=1]() # /home/lbachar/pytorch/test.py:9:0
#  %12 : Float(3, 4, strides=[4, 1], requires_grad=1, device=cpu) = aten::add[debug_name="MyCell/add1"](%20, %h, %11) # /home/lbachar/pytorch/test.py:9:0
#  %13 : Float(3, 4, strides=[4, 1], requires_grad=1, device=cpu) = aten::tanh(%12) # /home/lbachar/pytorch/test.py:9:0
#  %14 : (Float(3, 4, strides=[4, 1], requires_grad=1, device=cpu), Float(3, 4, strides=[4, 1], requires_grad=1, device=cpu)) = prim::TupleConstruct(%13, %13)
#  return (%14)
```
Adding debug_name attribute to a node will not affect future manipulations to the JIT graph often used in optimization passes  (i.e. subgraph rewriter pattern matching will ignore this attribute).
User can use this attribute to pass debug information to backend, similar to jit::ir::Value::unique_name_.

**Review by eellison — COMMENTED** [review]
I don't think we should do this. It's fine if you on the user side want to maintain a debug_name attribute on the IR (and you could that without using interned_strings - see `Symbol::attr("debug_name"))`. 

But adding this internally to the ir basically reproduces the problems of the using debugNames as anything but debug attributes. There's nothing preventing a graph having two nodes with the same debug name.. and it's not clear to me what exactly the contract this is with respect to collisions. 

If the intention is to aid subgraph matching we should discuss what else is possible here. Is it not enough to just do the matching as normal, and then filter through the results on the user-side such that debug names are equivalent ?

**Comment by hablb** [comment]
The comment above static Symbol attr(const std::string & s) in interned_strings.h:548 said not to use this function without adding the string as a literal.

The intention was to allow attaching debug name to node and show this information when printing jit graph, without interrupting subgraph matcher operation. Is it better if instead of using node attribute I will add debugName() method to node similar to Value::debugName() and print it when printing jit graph in Node::print()?

**Comment by hablb** [comment]
@eellison 
What are your thoughts on adding debugName() to node class?
No changes required to subgraph matcher and Node class would align with Value class.

**Comment by wconstab** [comment]
@hablb I think @eellison was proposing that you can do what you want to do without modifications to Node class at all.  You can just use `Symbol::attr("debug_name"))` from your user code.  If i understood that correctly, is there a problem with this approach?

**Pull request closed without merging**